src/daft-connect/src/display.rs

```diff
@@ -0,0 +1,326 @@
+    }
+    Ok(output)
+}
+
+// A helper function to print a field at a given level of indentation.
+// level=1 means a single " |-- " prefix, level=2 means
+// " |    |-- " and so on, mimicking Spark's indentation style.
+// A helper function to print a field at a given level of indentation.
+fn print_field(
+    w: &mut String,
+    field_name: &str,
+    dtype: &DataType,
+    nullable: bool,
+    level: usize,
+) -> eyre::Result<()> {
+    let indent = if level == 1 {
+        " |-- ".to_string()
+    } else {
+        format!(" |{}-- ", "    |".repeat(level - 1))
+    };
+
+    let dtype_str = type_to_string(dtype);
+    writeln!(
+        w,
+        "{}{}: {} (nullable = {})",
+        indent, field_name, dtype_str, nullable
+    )?;
+
+    if let DataType::Struct(fields) = dtype {
+        for field in fields {
+            print_field(w, &field.name, &field.dtype, true, level + 1)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn type_to_string(dtype: &DataType) -> String {
+    // We want a nice, human-readable type string.
+    // Spark generally prints something like "integer", "string", etc.
+    // We'll follow a similar style here:
+    match dtype {
+        DataType::Null => "null".to_string(),
+        DataType::Boolean => "boolean".to_string(),
+        DataType::Int8
+        | DataType::Int16
+        | DataType::Int32
+        | DataType::Int64
+        | DataType::UInt8
+        | DataType::UInt16
+        | DataType::UInt32
+        | DataType::UInt64 => "integer".to_string(), // Spark doesn't differentiate sizes
+        DataType::Float32 | DataType::Float64 => "double".to_string(), // Spark calls all floats double for printing
+        DataType::Decimal128(_, _) => "decimal".to_string(),
+        DataType::Timestamp(_, _) => "timestamp".to_string(),
+        DataType::Date => "date".to_string(),
+        DataType::Time(_) => "time".to_string(),
+        DataType::Duration(_) => "duration".to_string(),
+        DataType::Interval => "interval".to_string(),
+        DataType::Binary => "binary".to_string(),
+        DataType::FixedSizeBinary(_) => "fixed_size_binary".to_string(),
+        DataType::Utf8 => "string".to_string(),
+        DataType::FixedSizeList(_, _) => "array".to_string(), // Spark calls them arrays
+        DataType::List(_) => "array".to_string(),
+        DataType::Struct(_) => "struct".to_string(),
+        DataType::Map { .. } => "map".to_string(),
+        DataType::Extension(_, _, _) => "extension".to_string(),
+        DataType::Embedding(_, _) => "embedding".to_string(),
+        DataType::Image(_) => "image".to_string(),
+        DataType::FixedShapeImage(_, _, _) => "fixed_shape_image".to_string(),
+        DataType::Tensor(_) => "tensor".to_string(),
+        DataType::FixedShapeTensor(_, _) => "fixed_shape_tensor".to_string(),
+        DataType::SparseTensor(_) => "sparse_tensor".to_string(),
+        DataType::FixedShapeSparseTensor(_, _) => "fixed_shape_sparse_tensor".to_string(),
```

Comment:
i don't think these exist in spark (along with unsized ints). We should check if spark connect has a standard around extension or user defined types. If they don't I'd at least want something in the display to indicate that these are not native spark types, but in fact daft datatypes.

Reply:
hmmmmm https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/types/UserDefinedType.html

can people even define UDTs outside of Java? I don't see a pyspark example